File: LogisimPlus/LogisimPlus/logisim/tools/move/MoveGesture.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

using System;
using System.Collections.Generic;
using System.Threading;

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace logisim.tools.move
{

	using Circuit = logisim.circuit.Circuit;
	using Wire = logisim.circuit.Wire;
	using Component = logisim.comp.Component;
	using EndData = logisim.comp.EndData;
	using Direction = logisim.data.Direction;
	using Location = logisim.data.Location;

	public class MoveGesture
	{
		private MoveRequestListener listener;
		private Circuit circuit;
		private HashSet<Component> selected;

		[NonSerialized]
		private ISet<ConnectionData> connections;
		[NonSerialized]
		private AvoidanceMap initAvoid;
		private Dictionary<MoveRequest, MoveResult> cachedResults;

		public MoveGesture(MoveRequestListener listener, Circuit circuit, ICollection<Component> selected)
		{
			this.listener = listener;
			this.circuit = circuit;
			this.selected = new HashSet<Component>(selected);
			this.connections = null;
			this.initAvoid = null;
			this.cachedResults = new Dictionary<MoveRequest, MoveResult>();
		}

		internal virtual HashSet<Component> Selected
		{
			get
			{
				return selected;
			}
		}

		internal virtual AvoidanceMap FixedAvoidanceMap
		{
			get
			{
				AvoidanceMap ret = initAvoid;
				if (ret == null)
				{
					HashSet<Component> comps = new HashSet<Component>(circuit.NonWires);
					comps.addAll(circuit.Wires);
					comps.RemoveAll(selected);
					ret = AvoidanceMap.create(comps, 0, 0);
					initAvoid = ret;
				}
				return ret;
			}
		}

		internal virtual ISet<ConnectionData> Connections
		{
			get
			{
				ISet<ConnectionData> ret = connections;
				if (ret == null)
				{
					ret = computeConnections(circuit, selected);
					connections = ret;
				}
				return ret;
			}
		}

		public virtual MoveResult findResult(int dx, int dy)
		{
			MoveRequest request = new MoveRequest(this, dx, dy);
			lock (cachedResults)
			{
				return cachedResults[request];
			}
		}

		public virtual bool enqueueRequest(int dx, int dy)
		{
			MoveRequest request = new MoveRequest(this, dx, dy);
			lock (cachedResults)
			{
				object result = cachedResults[request];
				if (result == null)
				{
					ConnectorThread.enqueueRequest(request, false);
					return true;
				}
				else
				{
					return false;
				}
			}
		}

		public virtual MoveResult forceRequest(int dx, int dy)
		{
			MoveRequest request = new MoveRequest(this, dx, dy);
			ConnectorThread.enqueueRequest(request, true);
			lock (cachedResults)
			{
				object result = cachedResults[request];
				while (result == null)
				{
					try
					{
						Monitor.Wait(cachedResults);
					}
					catch (InterruptedException)
					{
						Thread.CurrentThread.Interrupt();
						return null;
					}
					result = cachedResults[request];
				}
				return (MoveResult) result;
			}
		}

		internal virtual void notifyResult(MoveRequest request, MoveResult result)
		{
			lock (cachedResults)
			{
				cachedResults[request] = result;
				Monitor.PulseAll(cachedResults);
			}
			if (listener != null)
			{
				listener.requestSatisfied(this, request.DeltaX, request.DeltaY);
			}
		}

		private static ISet<ConnectionData> computeConnections(Circuit circuit, ISet<Component> selected)
		{
			if (selected == null || selected.Count == 0)
			{
				return Collections.emptySet();
			}

			// first identify locations that might be connected
			ISet<Location> locs = new HashSet<Location>();
			foreach (Component comp in selected)
			{
				foreach (EndData end in comp.Ends)
				{
					locs.Add(end.Location);
				}
			}

			// now see which of them require connection
			ISet<ConnectionData> conns = new HashSet<ConnectionData>();
			foreach (Location loc in locs)
			{
				bool found = false;
				foreach (Component comp in circuit.getComponents(loc))
				{
					if (!selected.Contains(comp))
					{
						found = true;
						break;
					}
				}
				if (found)
				{
					IList<Wire> wirePath;
					Location wirePathStart;
					Wire lastOnPath = findWire(circuit, loc, selected, null);
					if (lastOnPath == null)
					{
						wirePath = Collections.emptyList();
						wirePathStart = loc;
					}
					else
					{
						wirePath = new List<Wire>();
						Location cur = loc;
						for (Wire w = lastOnPath; w != null; w = findWire(circuit, cur, selected, w))
						{
							wirePath.Add(w);
							cur = w.getOtherEnd(cur);
						}
						wirePath.Reverse();
						wirePathStart = cur;
					}

					Direction dir = null;
					if (lastOnPath != null)
					{
						Location other = lastOnPath.getOtherEnd(loc);
						int dx = loc.X - other.X;
						int dy = loc.Y - other.Y;
						if (Math.Abs(dx) > Math.Abs(dy))
						{
							dir = dx > 0 ? Direction.East : Direction.West;
						}
						else
						{
							dir = dy > 0 ? Direction.South : Direction.North;
						}
					}
					conns.Add(new ConnectionData(loc, dir, wirePath, wirePathStart));
				}
			}
			return conns;
		}

		private static Wire findWire(Circuit circ, Location loc, ISet<Component> ignore, Wire ignoreW)
		{
			Wire ret = null;
			foreach (Component comp in circ.getComponents(loc))
			{
				if (!ignore.Contains(comp) && comp != ignoreW)
				{
					if (ret == null && comp is Wire)
					{
						ret = (Wire) comp;
					}
					else
					{
						return null;
					}
				}
			}
			return ret;
		}

	}

}

File: logisim/app/src/main/java/logisim/tools/move/MoveGesture.java
/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

package logisim.tools.move;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import logisim.circuit.Circuit;
import logisim.circuit.Wire;
import logisim.comp.Component;
import logisim.comp.EndData;
import logisim.data.Direction;
import logisim.data.Location;

public class MoveGesture {
	private MoveRequestListener listener;
	private Circuit circuit;
	private HashSet<Component> selected;

	private transient Set<ConnectionData> connections;
	private transient AvoidanceMap initAvoid;
	private final HashMap<MoveRequest, MoveResult> cachedResults;

	public MoveGesture(MoveRequestListener listener, Circuit circuit, Collection<Component> selected) {
		this.listener = listener;
		this.circuit = circuit;
		this.selected = new HashSet<>(selected);
		connections = null;
		initAvoid = null;
		cachedResults = new HashMap<>();
	}

	HashSet<Component> getSelected() {
		return selected;
	}

	AvoidanceMap getFixedAvoidanceMap() {
		AvoidanceMap ret = initAvoid;
		if (ret == null) {
			HashSet<Component> comps = new HashSet<>(circuit.getNonWires());
			comps.addAll(circuit.getWires());
			comps.removeAll(selected);
			ret = AvoidanceMap.create(comps, 0, 0);
			initAvoid = ret;
		}
		return ret;
	}

	Set<ConnectionData> getConnections() {
		Set<ConnectionData> ret = connections;
		if (ret == null) {
			ret = computeConnections(circuit, selected);
			connections = ret;
		}
		return ret;
	}

	public MoveResult findResult(int dx, int dy) {
		MoveRequest request = new MoveRequest(this, dx, dy);
		synchronized (cachedResults) {
			return cachedResults.get(request);
		}
	}

	public boolean enqueueRequest(int dx, int dy) {
		MoveRequest request = new MoveRequest(this, dx, dy);
		synchronized (cachedResults) {
			Object result = cachedResults.get(request);
			if (result == null) {
				ConnectorThread.enqueueRequest(request, false);
				return true;
			} else return false;
		}
	}

	public MoveResult forceRequest(int dx, int dy) {
		MoveRequest request = new MoveRequest(this, dx, dy);
		ConnectorThread.enqueueRequest(request, true);
		synchronized (cachedResults) {
			MoveResult result = cachedResults.get(request);
			while (result == null) {
				try {
					cachedResults.wait();
				}
				catch (InterruptedException e) {
					Thread.currentThread().interrupt();
					return null;
				}
				result = cachedResults.get(request);
			}
			return result;
		}
	}

	void notifyResult(MoveRequest request, MoveResult result) {
		synchronized (cachedResults) {
			cachedResults.put(request, result);
			cachedResults.notifyAll();
		}
		if (listener != null) listener.requestSatisfied(this, request.getDeltaX(), request.getDeltaY());
	}

	private static Set<ConnectionData> computeConnections(Circuit circuit, Set<Component> selected) {
		if (selected == null || selected.isEmpty())
			return Collections.emptySet();

		// first identify locations that might be connected
		Set<Location> locs = new HashSet<>();
		for (Component comp : selected) for (EndData end : comp.getEnds()) locs.add(end.getLocation());

		// now see which of them require connection
		Set<ConnectionData> conns = new HashSet<>();
		for (Location loc : locs) {
			boolean found = false;
			for (Component comp : circuit.getComponents(loc))
				if (!selected.contains(comp)) {
					found = true;
					break;
				}
			if (found) {
				List<Wire> wirePath;
				Location wirePathStart;
				Wire lastOnPath = findWire(circuit, loc, selected, null);
				if (lastOnPath == null) {
					wirePath = Collections.emptyList();
					wirePathStart = loc;
				} else {
					wirePath = new ArrayList<>();
					Location cur = loc;
					for (Wire w = lastOnPath; w != null; w = findWire(circuit, cur, selected, w)) {
						wirePath.add(w);
						cur = w.getOtherEnd(cur);
					}
					Collections.reverse(wirePath);
					wirePathStart = cur;
				}

				Direction dir = null;
				if (lastOnPath != null) {
					Location other = lastOnPath.getOtherEnd(loc);
					int dx = loc.x() - other.x();
					int dy = loc.y() - other.y();
					if (Math.abs(dx) > Math.abs(dy)) dir = dx > 0 ? Direction.East : Direction.West;
					else dir = dy > 0 ? Direction.South : Direction.North;
				}
				conns.add(new ConnectionData(loc, dir, wirePath, wirePathStart));
			}
		}
		return conns;
	}

	private static Wire findWire(Circuit circ, Location loc, Set<Component> ignore, Wire ignoreW) {
		Wire ret = null;
		for (Component comp : circ.getComponents(loc))
			if (!ignore.contains(comp) && comp != ignoreW) if (ret == null && comp instanceof Wire) ret = (Wire) comp;
			else return null;
		return ret;
	}

}
